src/Filters/coefficients.jl

```diff
@@ -63,52 +83,101 @@ H(z) = \\frac{\\verb!b[1]! + \\ldots + \\verb!b[n]! z^{-n+1}}{\\verb!a[1]! + \\l
     end
-    PolynomialRatio{T}(Polynomial(reverse(b)), Polynomial(reverse(a)))
+    return promote(b / a[0], a / a[0])
+end
+function PolynomialRatio{D}(b::Polynomial, a::Polynomial) where {D}
+    b, a = _maybe_normalize(Val(D), b, a)
+    return PolynomialRatio{D,eltype(a)}(b, a)
 end
-PolynomialRatio(b::Union{T,Vector{T}}, a::Union{S,Vector{S}}) where {T<:Number,S<:Number} =
-    PolynomialRatio{promote_type(T,S)}(b, a)
 
-PolynomialRatio{T}(f::PolynomialRatio) where {T} = PolynomialRatio{T}(f.b, f.a)
-PolynomialRatio(f::PolynomialRatio{T}) where {T} = PolynomialRatio{T}(f)
+# The DSP convention for Laplace domain is highest power first. The Polynomials.jl
+# convention is lowest power first.
+_polyprep(D::Symbol, x) = D === :z ? x : reverse(x)
+_polyvar(D::Symbol) = D === :z ? Symbol("z^-1") : D
+PolynomialRatio{D,T}(b::Union{Number,Vector{<:Number}}, a::Union{Number,Vector{<:Number}}) where {D,T} =
+    PolynomialRatio{D,T}(Polynomial{T}(_polyprep(D, b), _polyvar(D)), Polynomial{T}(_polyprep(D, a), _polyvar(D)))
+PolynomialRatio{D}(b::Union{Number,Vector{<:Number}}, a::Union{Number,Vector{<:Number}}) where {D} =
+    PolynomialRatio{D}(Polynomial(_polyprep(D, b), _polyvar(D)), Polynomial(_polyprep(D, a), _polyvar(D)))
+
+PolynomialRatio{D,T}(f::PolynomialRatio{D}) where {D,T} =
+    PolynomialRatio{D,T}(convert(Polynomial{T}, f.b), convert(Polynomial{T}, f.a))
+PolynomialRatio{D}(f::PolynomialRatio{D,T}) where {D,T} = PolynomialRatio{D,T}(f)
 
-Base.promote_rule(::Type{PolynomialRatio{T}}, ::Type{PolynomialRatio{S}}) where {T,S} = PolynomialRatio{promote_type(T,S)}
+Base.promote_rule(::Type{PolynomialRatio{D,T}}, ::Type{PolynomialRatio{D,S}}) where {D,T,S} = PolynomialRatio{D,promote_type(T,S)}
 
-function PolynomialRatio{T}(f::ZeroPoleGain) where T<:Real
+function PolynomialRatio{:s,T}(f::ZeroPoleGain{:s}) where {T<:Real}
     b = f.k * fromroots(f.z)
     a = fromroots(f.p)
-    PolynomialRatio{T}(Polynomial(real(coeffs(b))), Polynomial(real(coeffs(a))))
+    return PolynomialRatio{:s,T}(Polynomial{T}(real(coeffs(b)), :s), Polynomial{T}(real(coeffs(a)), :s))
+end
+function PolynomialRatio{:z,T}(f::ZeroPoleGain{:z}) where {T<:Real}
+    b = real(f.k * coeffs(fromroots(f.z)))
+    a = real(coeffs(fromroots(f.p)))
+    if length(a) < length(b)
+        a = [a; zeros(eltype(a), length(b)-length(a))]
+    elseif length(b) < length(a)
+        b = [b; zeros(eltype(b), length(a)-length(b))]
+    end
+    return PolynomialRatio{:z,T}(reverse(b), reverse(a))
+end
+PolynomialRatio{D}(f::ZeroPoleGain{D,Z,P,K}) where {D,Z,P,K} =
+    PolynomialRatio{D,promote_type(real(Z),real(P),K)}(f)
+
+ZeroPoleGain{D,Z,P,K}(f::PolynomialRatio{D}) where {D,Z,P,K} =
+    ZeroPoleGain{D,Z,P,K}(ZeroPoleGain{D}(f))
+function ZeroPoleGain{:s}(f::PolynomialRatio{:s,T}) where {T}
+    z = roots(f.b)
+    p = roots(f.a)
+    # work around `roots` returning Any[] for polynomials of degree zero
```

Comment:
Ref. https://github.com/JuliaMath/Polynomials.jl/pull/286. If that gets accepted, we could remove this work-around if we bump the required `Polynomials` version.